### DISCUSSION TITLE
Adding worker node to K8s cluster created using Lima

### DISCUSSION THREAD

**Discussion by shivaveeru** [lima-vm/lima#1181] [Q&A]
I created K8s cluster using k8s.yaml. I have created worker node instance. But when I tried to join the worker node to cluster using kubeadm,I am receiving error" couldn't validate the identity of API server "https://127.0.0.1:6443" Has anyone tried the multinode cluster using Lima.Please let me know.

**Comment by afbjorklund** [comment]
This would need some better configuration of the URL in the config, obviously "localhost" won't work for a multi-node cluster...

`sed -e "s/${LIMA_CIDATA_SLIRP_IP_ADDRESS:-192.168.5.15}/127.0.0.1/" -i $KUBECONFIG`

But I think that it _does_ create certificates for both the IP and the name of the node, so maybe it will work if you change it ?

The kubeconfig was mostly intended to run from the host, with the 6443 being tunneled...

----

To create a multi-node cluster, it probably needs some "parameters" to the yaml : like if it is a control plane or a worker node

In order for the `join` command to work, it would also need to know 1) the address of the API server 2) the token from the CP

https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/create-cluster-kubeadm/#initializing-your-control-plane-node

https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/create-cluster-kubeadm/#join-nodes

The command to run should be in the `~/.lima/k8s/serial.log` file:

```
[  121.211642] cloud-init[2044]: Then you can join any number of worker nodes by running the following on each as root:
[  121.211734] cloud-init[2044]: kubeadm join 192.168.5.15:6443 --token z29xai.a676mmrpm5glpgsc \
[  121.213731] cloud-init[2044]: 	--discovery-token-ca-cert-hash sha256:73b89e14897e5e92c7082a1f7485ac865e68d3041da870d3b0ce201285773624
```

You need the **token**, the **url**, and the **hash** (discovery-token-ca-cert-hash)

This should be done, _instead_ of the whole the control plane system script...

----

EDIT: For k3s, the kubeconfig is in ` /etc/rancher/k3s/k3s.yaml`

And the token is stored in `/var/lib/rancher/k3s/server/node-token`

https://docs.k3s.io/quick-start

If using k3s.yaml instead, that is.

**Reply by afbjorklund** [reply]
Added some placeholder code in PR #1183, so that you only need to fill in a one-liner with the needed k3s/k8s information:

`export K3S_TOKEN=xxx K3S_URL=https://server-url:6443`

`export K8S_URL=server-url:6443 K8S_TOKEN=xxx K8S_HASH=sha256:<hash>`

Maybe not as nice as the parameters (like for e.g. mounts or containerd), but a little better perhaps. Still need the token...

----

Note that the above example with `192.168.5.15` does **not** work! They all have that exact same IP address...

See https://github.com/lima-vm/lima/blob/master/docs/network.md

It will need some more detailed documentation, on how to start a multi-node cluster including the networking.

Possibly also a "IP" feature to have the apiserver listen on the `eth1` interface, if the slirp is still on the `eth0` ?


--apiserver-advertise-address string
  | The IP address the API Server will advertise it's listening on. If not set the default network interface will be used.
--apiserver-bind-port int32     Default: 6443
  | Port for the API Server to bind to.
--apiserver-cert-extra-sans strings
  | Optional extra Subject Alternative Names (SANs) to use for the API Server serving certificate. Can be both IP addresses and DNS names.

**Reply by shivaveeru** [reply]
One thing. I noticed is the master node and worker instance has same ip i.e 192.168.5.15. This is ambiguous right. Is there anyway I can specify particular ip address to worker node? Please refer to (https://github.com/lima-vm/lima/discussions/480#discussion-3742273) for shared network. Now it makes sense after your latest comment

**Reply by afbjorklund** [reply]
Yeah, I forgot to mention the network requirement (at first)

Would like a network solution that also works on developer machines, with Linux

**Reply by afbjorklund** [reply]
Added an example in PR #1185  of two nodes with a mutual network, left todo is to get the cluster to run on `eth1` instead of `eth0`.

```yaml
    kind: InitConfiguration
    apiVersion: kubeadm.k8s.io/v1beta3
    localAPIEndpoint:
      advertiseAddress: 10.0.2.100
      bindPort: 6443
    nodeRegistration:
      criSocket: unix:///run/containerd/containerd.sock
```

And edit the kube-flannel-ds (DaemonSet), to add the `--iface=eth1` arg to `flanneld`.

worker: `K8S_URL=https://10.0.2.100:6443` (when running the modified k8s.yaml example, from PR #1183)

**Reply by afbjorklund** [reply]
This probably also needs to add `--node-ip` to the kubelet, or there is bound to be confusion...

```
NAME              STATUS   ROLES           AGE   VERSION
lima-k8s-master   Ready    control-plane   49m   v1.25.4
lima-k8s-minion   Ready    <none>          41m   v1.25.4
```

```
NAMESPACE      NAME                                      READY   STATUS    RESTARTS   AGE     IP             NODE              NOMINATED NODE   READINESS GATES
kube-flannel   kube-flannel-ds-fldh6                     1/1     Running   0          3m57s   192.168.5.15   lima-k8s-master   <none>           <none>
kube-flannel   kube-flannel-ds-pjdlg                     1/1     Running   0          3m52s   192.168.5.15   lima-k8s-minion   <none>           <none>
kube-system    coredns-565d847f94-clqjd                  1/1     Running   0          52m     10.244.0.2     lima-k8s-master   <none>           <none>
kube-system    coredns-565d847f94-w476b                  1/1     Running   0          52m     10.244.0.3     lima-k8s-master   <none>           <none>
kube-system    etcd-lima-k8s-master                      1/1     Running   0          52m     192.168.5.15   lima-k8s-master   <none>           <none>
kube-system    kube-apiserver-lima-k8s-master            1/1     Running   0          52m     192.168.5.15   lima-k8s-master   <none>           <none>
kube-system    kube-controller-manager-lima-k8s-master   1/1     Running   0          52m     192.168.5.15   lima-k8s-master   <none>           <none>
kube-system    kube-proxy-ct52b                          1/1     Running   0          52m     192.168.5.15   lima-k8s-master   <none>           <none>
kube-system    kube-proxy-dwgl6                          1/1     Running   0          45m     192.168.5.15   lima-k8s-minion   <none>           <none>
kube-system    kube-scheduler-lima-k8s-master            1/1     Running   0          52m     192.168.5.15   lima-k8s-master   <none>           <none>
```

There doesn't seem to be any straightforward way of doing so with `kubeadm`, unfortunately ? 😢

* https://github.com/kubernetes/kubeadm/issues/203 (from 2017)

Only `/var/lib/kubelet/kubeadm-flags.env` (/etc/systemd/system/kubelet.service.d/10-kubeadm.conf)

And then restarting the kubelet service (after adding --node-ip), to use the new IP addresses for the nodes.

```console
$ limactl ls
NAME                STATUS     SSH                ARCH      CPUS    MEMORY    DISK      DIR
k8s-master          Running    127.0.0.1:40199    x86_64    4       4GiB      100GiB    /home/anders/.lima/k8s-master
k8s-minion          Running    127.0.0.1:41883    x86_64    4       4GiB      100GiB    /home/anders/.lima/k8s-minion
$ kubectl get nodes -o wide
NAME              STATUS   ROLES           AGE    VERSION   INTERNAL-IP   EXTERNAL-IP   OS-IMAGE             KERNEL-VERSION      CONTAINER-RUNTIME
lima-k8s-master   Ready    control-plane   118m   v1.25.4   10.0.2.100    <none>        Ubuntu 22.04.1 LTS   5.15.0-52-generic   containerd://1.6.8
lima-k8s-minion   Ready    <none>          110m   v1.25.4   10.0.2.101    <none>        Ubuntu 22.04.1 LTS   5.15.0-52-generic   containerd://1.6.8
```

**Reply by jandubois** [reply]
For `k3s` you can specify the `--flannel-iface eth1` option.

**Reply by afbjorklund** [reply]
> For `k3s` you can specify the `--flannel-iface eth1` option.

It has `--node-ip` as well, and of course it also has get.k3s.io

Kubernetes has something user-friendly like: (guess you could pipe it)

`sed -i '/--ip-masq/ i\ \ \ \ \ \ \ \ - --iface=eth1' kube-flannel.yml`

----

https://raw.githubusercontent.com/flannel-io/flannel/v0.20.1/Documentation/kube-flannel.yml

```diff
@@ -154,6 +154,7 @@
         command:
         - /opt/bin/flanneld
         args:
+        - --iface=eth1
         - --ip-masq
         - --kube-subnet-mgr
         resources:

```

**Reply by louhisuo** [reply]
I am working with multi-node kubernetes environments and therefore I have been looking how to quickly spinup different labs on a macbook with apple silicon.

When looking Lima here, it has promise but one of the biggest issue I see with Lima at the moment is inability to set Static IPs for control plane and worker nodes. I can get multi-node Kubernetes up and running by using user-v2 network but as I do not have way to set static IPs to all nodes. I am not able to configure in kubeadm configuration (v1beta3) things like 'advertiseAddress' and controlPlaneEndpoint which assumes static IP.

My current understanding is that I could even use shared network mode for multi-node Kubernetes cluster if configuring static IPs for VMs were supported. Shared network mode would also make possible to build HA control configurations where cluster would be accessable from host as well.

I am well aware that KiND can be also used for this and things are a bit easier with K3s but those solutions has issues as well. And when working with a bit more advanced / forward leaning technologies like Cilium CNI, eBFP, Service meshes, Gateway API etc. we really need static IP support to able to work Kubernetes feature evaluations.

**Reply by afbjorklund** [reply]
For my other (non-Lima) clusters I set the MAC address (for VM and BM*), and then have the DHCP server map them to the static IP.

With Lima, I just hacked the static IP for the built-in "socket" network. There is probably some more fancy way to do it with gvproxy

* https://github.com/lima-vm/lima/pull/1185

* https://github.com/lima-vm/lima/pull/1383

----

\* BM = 🐻 🤘

**Reply by louhisuo** [reply]
In non-Lima environments I typically rely on static IPs which are configured either in Coud-init or Ignition (depending which OS requires/prefers. For example with Multipass on macOS have used Cloud-init. In rare cases, I may rely on DHCP Client ID or MAC address with DHCP server to provide stable predictable IP.

Regarding PoCs you made are promising direction, however my understanding is that user-v2 networks (#1383) will not support host - VM communication, only VM-VM communication. Having said so, it would be useful to have stable, predictable IPs also for user-v2 networks, especially if it will be possible to provide this IP via limactl create/start (I am looking here to avoid having dedicated lima.yaml for every Kubernetes node, as their configurations are widely identical).

Ultimately some cloud-init or Vagrant type of way to provide stable predictable IP is something I would personally like, with possibility to overide lima.yaml value via limactl create/start.

**Reply by afbjorklund** [reply]
For the earlier kubernetes clusters, we used **two** networks: one `eth0` for host communication, and one `eth1` for VM-to-VM

The biggest hassle was convincing Kubernetes to start the server on the _second_ interface, as shown in discussion above...

**Reply by louhisuo** [reply]
Yes, I was reading that and I am trying to understand why you have issues to get kube-api server to consume second interface.

I did quick test and I had no issues to get kube-api server to use second interface in VM as long I assigned same IP address which was allocated to the second interface (lima1) to `advertiseAddress` field in kubeadm `InitConfiguration` section. This ensured that all configurations and certifcates got generated correctly and I just simply pasted `kubeadm join` to second VM and got two node cluster (1x CP and 1x WN) running without issues.

**Reply by afbjorklund** [reply]
Not sure if was the kube-apiserver, I think it was how the node presented itself. Anyway, there were workarounds even so.

**Reply by louhisuo** [reply]
Are referring that --node-ip flag (kubelet configuration, if I recall correctly)? If so I would need to look which cases it would really need to be set separatelly?

Regardless, I did not need to do anything else than just to set `advertiseAddress` to same address which got assigned to `lima1` interface to get this working. See example below. `192.168.105.49`was same IP which is also assigned IP to `lima1`interface in test cluster I spinned up.

kubeadm config
```
apiVersion: kubeadm.k8s.io/v1beta3
kind: InitConfiguration
localAPIEndpoint:
  advertiseAddress: "192.168.105.49"
  bindPort: 6443
```
`ip address` command snippet taken from control plane node
```
2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc fq_codel state UP group default qlen 1000
    link/ether 52:55:55:c9:a5:06 brd ff:ff:ff:ff:ff:ff
    altname enp0s2
    inet 192.168.5.5/24 metric 100 brd 192.168.5.255 scope global dynamic eth0
       valid_lft 3035sec preferred_lft 3035sec
    inet6 fe80::5055:55ff:fec9:a506/64 scope link 
       valid_lft forever preferred_lft forever
3: lima1: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc fq_codel state UP group default qlen 1000
    link/ether 52:55:55:b3:d6:4c brd ff:ff:ff:ff:ff:ff
    altname enp0s3
    inet 192.168.105.49/24 metric 100 brd 192.168.105.255 scope global dynamic lima1
       valid_lft 82234sec preferred_lft 82234sec
    inet6 fd18:e25f:4ff4:a24f:5055:55ff:feb3:d64c/64 scope global dynamic mngtmpaddr noprefixroute 
       valid_lft 2591962sec preferred_lft 604762sec
    inet6 fe80::5055:55ff:feb3:d64c/64 scope link 
       valid_lft forever preferred_lft forever
```

**Reply by afbjorklund** [reply]
As long as your nodes and your pods get different IPs, there should be no issue. You can check with `-o wide`.

**Reply by louhisuo** [reply]
They got, checked this earlier earlier. But I just noticed that it is `eth0` IP assigned as internal IP for Kubernetes nodes so will need to check this why and if it matters.
```
$ kubectl get nodes -o wide
NAME          STATUS   ROLES           AGE   VERSION   INTERNAL-IP   EXTERNAL-IP   OS-IMAGE             KERNEL-VERSION      CONTAINER-RUNTIME
lima-test     Ready    control-plane   94m   v1.28.1   192.168.5.5   <none>        Ubuntu 22.04.3 LTS   5.15.0-82-generic   containerd://1.7.3
lima-test-2   Ready    <none>          91m   v1.28.1   192.168.5.6   <none>        Ubuntu 22.04.3 LTS   5.15.0-82-generic   containerd://1.7.3

$ kubectl get pods -A -o wide
NAMESPACE     NAME                                READY   STATUS    RESTARTS   AGE   IP            NODE          NOMINATED NODE   READINESS GATES
kube-system   cilium-kv8ln                        1/1     Running   0          84m   192.168.5.6   lima-test-2   <none>           <none>
kube-system   cilium-operator-788c4f69bc-6nstc    1/1     Running   0          84m   192.168.5.6   lima-test-2   <none>           <none>
kube-system   cilium-zfrzt                        1/1     Running   0          84m   192.168.5.5   lima-test     <none>           <none>
kube-system   coredns-5dd5756b68-cv8gl            1/1     Running   0          89m   10.0.0.198    lima-test     <none>           <none>
kube-system   coredns-5dd5756b68-khkqb            1/1     Running   0          89m   10.0.0.163    lima-test     <none>           <none>
kube-system   etcd-lima-test                      1/1     Running   0          89m   192.168.5.5   lima-test     <none>           <none>
kube-system   kube-apiserver-lima-test            1/1     Running   0          89m   192.168.5.5   lima-test     <none>           <none>
kube-system   kube-controller-manager-lima-test   1/1     Running   0          89m   192.168.5.5   lima-test     <none>           <none>
kube-system   kube-proxy-2tqlf                    1/1     Running   0          89m   192.168.5.5   lima-test     <none>           <none>
kube-system   kube-proxy-wfdm9                    1/1     Running   0          86m   192.168.5.6   lima-test-2   <none>           <none>
kube-system   kube-scheduler-lima-test            1/1     Running   0          89m   192.168.5.5   lima-test     <none>           <none>
```

**Reply by louhisuo** [reply]
Looks like `kubelet` uses default IP for a node which is IP for the first interface if `--node-ip` is not set. This is the reason why I see user-v2 network addresses as node `INTERNAL-IP`s. If I remember correctly kubeadm config has a separate section to configure `kubelet` so I can probably manipulate `--node-ip` flag value in that way.

**Reply by louhisuo** [reply]
Did some further testing and succeeded to setup 1x Control Plane and 3x Workers Kubernetes cluster with `kubeadm` with fairly little efforts (and no hacks). 

I used `Shared` network, instead of `user-v2` network in Lima because for me `Host-VM` communication is also important. I set `advertiseAddress`, `node-ip` and `controlPlaneEndpoint` fields in kubeadm init config (`InitConfiguration` and `ClusterConfiguration`) file to IP assigned to `lima0`interface in Control Plane node. I had also kubeadm join config file where I set `apiServerEndpoint` to point `controlPlaneEndpoint` in init config and `node-ip` to IP assigned to `lima0` interface for a Worker node. Then instead of passing `kubeadm join`command around, I have pre-configured `discovery.bootstrapToken` to simplify the process for workers to join the cluster. 

I used Kubernetes 1.28.1, Cilium 1.14.1 (CNI), MetalLB (L2 loadbalancer), Ingress-NGINX (L7 Ingress Controller) and an Echo-Server app to confirm quickly that things work. I did not try Control Plane HA setup with `kube-vip` yet, neither tried Cilium without `kube-proxy` but those will hopefully work as well.

So in summary, it is possible to setup multi-node K8s cluster using current Lima 0.17.2 release (kudos for Lima team). No hacks needed. Lima is used to provision VMs and `kubeadm` boot up a cluster after VMs are up and running. This process could be further streamlined if Lima starts supporting static or pre-configurable preditable IPs for its `Shared` and `user-v2` networks somewhere in future. My personal preference here would be `Shared`network as it gives `Host-VM` connectivity in addition to `VM-VM` connectivity enabling e.g. Kubernetes focused portable demos as an example use case. For development, people probably will be happy with `VM-VM`connectivity provided by `user-v2`network.

**Reply by louhisuo** [reply]
Per advice in [#1785](https://github.com/lima-vm/lima/issues/1785#issuecomment-1720980302), it was possible to make IP addresses deterministic when using `shared` network. 

To keep `lima.yaml` as common template for all VMs created, I used below command syntax to (also `limactl create` works equally well) to create VMs.
```
limactl start --set='.networks[].macAddress = "52:55:55:12:34:56"' --name test test.yaml
```
Works like sharm. It is now possible to prepare `kubeadm init`and `kubeadm join`configs for a cluster in advance, as IP addresses can be planned in advance, making it possible to create/tear down multi-node Kubernetes clusters quickly and with little manual efforts.
